src/openstack_cloud/openstack_manager.py

```diff
@@ -0,0 +1,378 @@
+    # I could not reproduce it. Therefore, no catch here.
+    return openstack.connect(cloud_name)
+
+
+def list_projects(cloud_config: dict[str, dict]) -> list[Project]:
+    """List all projects in the OpenStack cloud.
+
+    The purpose of the method is just to try out openstack integration and
+    it may be removed in the future.
+
+    It currently returns objects directly from the sdk,
+    which may not be ideal (mapping to domain objects may be preferable).
+
+    Returns:
+        A list of projects.
+    """
+    conn = _create_connection(cloud_config)
+    try:
+        projects = conn.list_projects()
+        logger.debug("OpenStack connection successful.")
+        logger.debug("Projects: %s", projects)
+        # pylint thinks this isn't an exception
+    except keystoneauth1.exceptions.http.Unauthorized as exc:
+        raise OpenStackUnauthorizedError(  # pylint: disable=bad-exception-cause
+            "Unauthorized to connect to OpenStack."
+        ) from exc
+
+    return projects
+
+
+def _generate_docker_proxy_unit_file(proxies: Optional[ProxyConfig] = None) -> str:
+    """Generate docker proxy systemd unit file.
+
+    Args:
+        proxies: HTTP proxy settings.
+
+    Returns:
+        Contents of systemd-docker-proxy unit file.
+    """
+    environment = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+    return environment.get_template("systemd-docker-proxy.j2").render(proxies=proxies)
+
+
+def _generate_docker_client_proxy_config_json(http_proxy: str, https_proxy: str, no_proxy: str):
+    """Generate proxy config.json for docker client.
+
+    Args:
+        http_proxy: HTTP proxy URL.
+        https_proxy: HTTPS proxy URL.
+        no_proxy: URLs to not proxy through.
+
+    Returns:
+        Contents of docker config.json file.
+    """
+    return json.dumps(
+        {
+            "proxies": {
+                "default": {
+                    key: value
+                    for key, value in (
+                        ("httpProxy", http_proxy),
+                        ("httpsProxy", https_proxy),
+                        ("noProxy", no_proxy),
+                    )
+                    if value
+                }
+            }
+        }
+    )
+
+
+def _build_image_command(
+    runner_info: RunnerApplication, proxies: Optional[ProxyConfig] = None
+) -> list[str]:
+    """Get command for building runner image.
+
+    Args:
+        runner_info: The runner application to fetch runner tar download url.
+        proxies: HTTP proxy settings.
+
+    Returns:
+        Command to execute to build runner image.
+    """
+    docker_proxy_service_conf_content = _generate_docker_proxy_unit_file(proxies=proxies)
+
+    http_proxy = str(proxies.http) if (proxies and proxies.http) else ""
+    https_proxy = str(proxies.https) if (proxies and proxies.https) else ""
+    no_proxy = str(proxies.no_proxy) if (proxies and proxies.no_proxy) else ""
```

Comment:
Would it be good to extract the None check for `proxies` out? It does increase the nesting by one, but it would avoid repeated None check for line 147 - 149.

Reply:
Refactored, thank you!